
<p style="font-size:21px; font-family:'Times New Roman';">
Read report dates data 
</p>

In [1]:
import pandas as pd
import numpy as np
csv_file_path = '/Users/balmeru/Downloads/QQQQ.csv'
df = pd.read_csv(csv_file_path)


In [2]:
df.head()

GVKEY    datadate  fyearq  fqtr  fyr indfmt consol popsrc datafmt   tic  \
0   1000  1975-06-30    1975     2   12   INDL      C      D     STD  AE.2   
1   1000  1975-09-30    1975     3   12   INDL      C      D     STD  AE.2   
2   1000  1975-12-31    1975     4   12   INDL      C      D     STD  AE.2   
3   1000  1976-03-31    1976     1   12   INDL      C      D     STD  AE.2   
4   1000  1976-06-30    1976     2   12   INDL      C      D     STD  AE.2   

   ...    ibq    lseq     ltq  pstkq teqq  txdbq   dvpy  costat  mkvaltq  \
0  ...  0.875     NaN     NaN  0.414  NaN    NaN    NaN       I      NaN   
1  ...  0.613     NaN     NaN  0.414  NaN    NaN    NaN       I      NaN   
2  ...  0.606  23.905  11.596  2.069  NaN    NaN    NaN       I      NaN   
3  ...  0.726  32.531  19.469  2.069  NaN    NaN  0.031       I      NaN   
4  ...  1.603  35.157  20.287  2.069  NaN    NaN    NaN       I      NaN   

   prccq  
0  2.875  
1  4.000  
2  4.375  
3  6.750  
4  7.250  

[5 rows x 29 columns]

<p style="font-size:21px; font-family:'Times New Roman';">
Let's address fiscal year changes first </p>


In [3]:
df['market_cap'] = df['prccq'] / df['ajexq'] * df['cshoq']


In [4]:
print(df[['prccq', 'ajexq', 'cshoq', 'market_cap']].head())


   prccq  ajexq  cshoq  market_cap
0  2.875    1.0  2.092     6.01450
1  4.000    1.0  2.094     8.37600
2  4.375    1.0  2.098     9.17875
3  6.750    1.0  2.096    14.14800
4  7.250    1.0  2.163    15.68175


In [5]:
non_empty_market_cap_count = df['market_cap'].notna().sum()

print(f"Number of non-empty market_cap rows: {non_empty_market_cap_count}")


Number of non-empty market_cap rows: 1014935


In [6]:
num_rows = df.shape[0]

print(f"The number of rows in the CSV file is: {num_rows}")


The number of rows in the CSV file is: 1085915


In [7]:
print(df[['prccq', 'ajexq', 'cshoq', 'market_cap', 'mkvaltq']].head())


   prccq  ajexq  cshoq  market_cap  mkvaltq
0  2.875    1.0  2.092     6.01450      NaN
1  4.000    1.0  2.094     8.37600      NaN
2  4.375    1.0  2.098     9.17875      NaN
3  6.750    1.0  2.096    14.14800      NaN
4  7.250    1.0  2.163    15.68175      NaN


In [8]:
non_empty_market_val_count = df['mkvaltq'].notna().sum()

print(f"Number of non-empty market_cap rows: {non_empty_market_val_count}")

Number of non-empty market_cap rows: 287568


In [9]:
non_empty_count = df[['mkvaltq', 'market_cap']].notna().any(axis=1).sum()
print(f"Number of rows with non-empty mkvaltq or market_cap: {non_empty_count}")





Number of rows with non-empty mkvaltq or market_cap: 1015011


In [10]:
non_empty_rows = df[df['mkvaltq'].notna() & df['market_cap'].notna()]
print(non_empty_rows)


          GVKEY    datadate  fyearq  fqtr  fyr indfmt consol popsrc datafmt  \
170        1004  2006-08-31    2006     1    5   INDL      C      D     STD   
171        1004  2006-11-30    2006     2    5   INDL      C      D     STD   
172        1004  2007-02-28    2006     3    5   INDL      C      D     STD   
173        1004  2007-05-31    2006     4    5   INDL      C      D     STD   
174        1004  2007-08-31    2007     1    5   INDL      C      D     STD   
...         ...         ...     ...   ...  ...    ...    ...    ...     ...   
1085910  345556  2021-03-31    2021     1   12   INDL      C      D     STD   
1085911  345920  2020-12-31    2020     4   12   INDL      C      D     STD   
1085912  345920  2021-03-31    2021     1   12   INDL      C      D     STD   
1085913  345980  2020-12-31    2020     4   12   INDL      C      D     STD   
1085914  345980  2021-03-31    2021     1   12   INDL      C      D     STD   

          tic  ...      lseq       ltq pstkq      t

In [11]:
decimal_places = 2  # Adjust this as needed
rounded_mkvaltq = df['mkvaltq'].round(decimal_places)
rounded_market_cap = df['market_cap'].round(decimal_places)
identical_count = (rounded_mkvaltq == rounded_market_cap).sum()

print(f"Number of identical or rounded values between mkvaltq and market_cap: {identical_count}")


Number of identical or rounded values between mkvaltq and market_cap: 223395


 As you see, the number of rows for calculated **market cap** is 1,014,935 out of 1,085,915 entries available.

 While the given **mkvaltq** (Market Value Total) variable in the data itself has only 287,568 rows.

 As you see from the printed rows, market value and calculated market cap can be same or different; they *differ* more than they have identical values.


#### To be consistent, we can proceed with calculated market cap now

In [12]:
# Calculate upper and lower bounds
df['lower_bound'] = 0.8 * df['market_cap']
df['upper_bound'] = 1.2 * df['market_cap']
print(df[['market_cap', 'lower_bound', 'upper_bound']])


            market_cap    lower_bound    upper_bound
0             6.014500       4.811600       7.217400
1             8.376000       6.700800      10.051200
2             9.178750       7.343000      11.014500
3            14.148000      11.318400      16.977600
4            15.681750      12.545400      18.818100
...                ...            ...            ...
1085910      98.553000      78.842400     118.263600
1085911    1761.430000    1409.144000    2113.716000
1085912    2049.070400    1639.256320    2458.884480
1085913  321206.432121  256965.145697  385447.718545
1085914  292932.029293  234345.623435  351518.435152

[1085915 rows x 3 columns]


In [13]:
df['quarterly_report'] = (df['fqtr'].isin([1, 2, 3])).astype(int)
df['annual_report'] = (df['fqtr'] == 4).astype(int)
print(df[['fqtr', 'quarterly_report', 'annual_report']])



         fqtr  quarterly_report  annual_report
0           2                 1              0
1           3                 1              0
2           4                 0              1
3           1                 1              0
4           2                 1              0
...       ...               ...            ...
1085910     1                 1              0
1085911     4                 0              1
1085912     1                 1              0
1085913     4                 0              1
1085914     1                 1              0

[1085915 rows x 3 columns]


In [14]:
df.tail()

GVKEY    datadate  fyearq  fqtr  fyr indfmt consol popsrc datafmt  \
1085910  345556  2021-03-31    2021     1   12   INDL      C      D     STD   
1085911  345920  2020-12-31    2020     4   12   INDL      C      D     STD   
1085912  345920  2021-03-31    2021     1   12   INDL      C      D     STD   
1085913  345980  2020-12-31    2020     4   12   INDL      C      D     STD   
1085914  345980  2021-03-31    2021     1   12   INDL      C      D     STD   

          tic  ...  txdbq dvpy costat     mkvaltq  prccq     market_cap  \
1085910  FSTX  ...  0.576  0.0      I     98.5530  10.83      98.553000   
1085911  HYFM  ...  0.000  0.0      A   1761.4300  52.58    1761.430000   
1085912  HYFM  ...  0.000  0.0      A   2049.0704  60.32    2049.070400   
1085913  LOGC  ...    NaN  0.0      A  10706.8800  18.24  321206.432121   
1085914  LOGC  ...    NaN  0.0      A   9764.4000  15.80  292932.029293   

           lower_bound    upper_bound  quarterly_report  annual_report  
1085910      78.842400     118.263600                 1              0  
1085911    1409.144000    2113.716000                 0              1  
1085912    1639.256320    2458.884480                 1              0  
1085913  256965.145697  385447.718545                 0              1  
1085914  234345.623435  351518.435152                 1              0  

[5 rows x 34 columns]

In [15]:


# Convert rdq to datetime if it is not already
df['rdq'] = pd.to_datetime(df['rdq'], errors='coerce')
df['fyearq'] = pd.to_numeric(df['fyearq'], errors='coerce')
df['fqtr'] = pd.to_numeric(df['fqtr'], errors='coerce')

grouped = df.groupby('tic', group_keys=False)

# Function to calculate consecutive differences
def calculate_diff(group):
    # Sort the data within each group by the fiscal year and quarter in ascending order
    group = group.sort_values(['fyearq', 'fqtr'])
    
    # Create a shifted column for the previous rdq
    group['prev_rdq'] = group['rdq'].shift(1)
    group['prev_fqtr'] = group['fqtr'].shift(1)
    group['prev_fyearq'] = group['fyearq'].shift(1)
    
    # Calculate the differences
    group['diff'] = group['rdq'] - group['prev_rdq']
    
    # Remove differences where quarters and years are not consecutive
    mask = (group['fqtr'] == 1) & (group['prev_fqtr'] == 4) & (group['fyearq'] == group['prev_fyearq'] + 1)
    mask |= (group['fqtr'] > 1) & (group['fqtr'] == group['prev_fqtr'] + 1) & (group['fyearq'] == group['prev_fyearq'])
    
    group['diff'] = group['diff'].where(mask)
    
    # Drop helper columns
    group = group.drop(columns=['prev_rdq', 'prev_fqtr', 'prev_fyearq'])
    
    return group

# Apply the function to each ticker group directly in the original DataFrame
df = grouped.apply(calculate_diff)

# Reset index to make it easier to read
df = df.reset_index(drop=True)

# Display the DataFrame with the computed differences
print(df[['tic', 'rdq', 'fyearq', 'fqtr', 'diff']].tail(30))


/var/folders/ds/rfyxvvb945vdt8znlkx31w1m00cj2y/T/ipykernel_14095/771304066.py:33: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df = grouped.apply(calculate_diff)


          tic        rdq  fyearq  fqtr     diff
1085885    ZZ 2012-01-18    2011     4 113 days
1085886    ZZ 2012-03-27    2012     1  69 days
1085887    ZZ 2012-06-26    2012     2  91 days
1085888    ZZ 2012-09-27    2012     3  93 days
1085889    ZZ 2013-01-23    2012     4 118 days
1085890  ZZZ1        NaT    1994     2      NaT
1085891  ZZZ1 1994-10-19    1994     3      NaT
1085892  ZZZ1 1995-01-17    1994     4  90 days
1085893  ZZZ1 1995-04-19    1995     1  92 days
1085894  ZZZ1 1995-07-20    1995     2  92 days
1085895  ZZZ1 1995-10-19    1995     3  91 days
1085896  ZZZ1 1996-01-25    1995     4  98 days
1085897  ZZZ1 1996-04-17    1996     1  83 days
1085898  ZZZ1 1996-07-17    1996     2  91 days
1085899  ZZZ1 1996-10-17    1996     3  92 days
1085900  ZZZ1 1997-01-28    1996     4 103 days
1085901  ZZZ1 1997-04-25    1997     1  87 days
1085902  ZZZ1 1997-07-23    1997     2  89 days
1085903  ZZZ1 1997-10-22    1997     3  91 days
1085904  ZZZ1 1998-01-28    1997     4  

#### After calculating the difference between consecutive reports, now we want to identify tickers for which fiscal year change happened.
#### We want to set the dummy variable to 1 at the specific points where the fiscal year end month (fyr) changes for each ticker

In [16]:
df = df.sort_values(by=['tic', 'fyr'])

# Find where tickers change
ticker_change = df['tic'] != df['tic'].shift(1)

# Create the dummy variable
df['fyr_change_dummy'] = ((df['fyr'] != df['fyr'].shift(1)) & (~ticker_change)).astype(int)

# Fill the first occurrence of each ticker group with 0
df.loc[df.groupby('tic').head(1).index, 'fyr_change_dummy'] = 0
print(df[['tic', 'rdq', 'fyr', 'fyearq', 'fqtr', 'quarterly_report', 'annual_report', 'fyr_change_dummy', 'market_cap', 'lower_bound', 'upper_bound', 'diff']].head())


      tic        rdq  fyr  fyearq  fqtr  quarterly_report  annual_report  \
58  0015B 1998-09-23    5    1998     1                 1              0   
59  0015B 1998-12-23    5    1998     2                 1              0   
60  0015B 1999-03-17    5    1998     3                 1              0   
61  0015B        NaT    5    1998     4                 0              1   
62  0015B 1999-09-15    5    1999     1                 1              0   

    fyr_change_dummy  market_cap  lower_bound  upper_bound     diff  
58                 0  991.896313    793.51705  1190.275575 378 days  
59                 0  719.272500    575.41800   863.127000  91 days  
60                 0  598.121000    478.49680   717.745200  84 days  
61                 0  783.000000    626.40000   939.600000      NaT  
62                 0  765.699000    612.55920   918.838800      NaT  


In [17]:
num_changes = (df['fyr_change_dummy'] == 1).sum()

print("Number of changes:", num_changes)

Number of changes: 2070


In [18]:
filtered_df = df[df['fyr_change_dummy'] == 1]
print(filtered_df[['tic', 'rdq', 'fyr', 'fyearq', 'fqtr', 'quarterly_report', 'annual_report', 'fyr_change_dummy', 'market_cap', 'lower_bound', 'upper_bound', 'diff']].head())



        tic        rdq  fyr  fyearq  fqtr  quarterly_report  annual_report  \
22    0015B 1989-03-10    6    1989     1                 1              0   
0     0015B        NaT   10    1983     3                 1              0   
625   0097A        NaT   12    1986     4                 0              1   
1055  0147A 1991-07-31   12    1991     2                 1              0   
1206  0161A 1999-04-28    6    1999     3                 1              0   

      fyr_change_dummy  market_cap  lower_bound  upper_bound    diff  
22                   1         NaN          NaN          NaN 43 days  
0                    1   98.222171    78.577737   117.866605     NaT  
625                  1   19.499994    15.599995    23.399993     NaT  
1055                 1   42.588715    34.070972    51.106458     NaT  
1206                 1  687.925938   550.340750   825.511125     NaT  


#### Split difference in report dates for quarterly reports and annual report.
#### For quarterly reports: Calculate the difference between each quarter and the previous one.
#### For annual reports (4th quarter): Calculate the difference from the 3rd quarter of the same year.

#### Let's split the distance for quarterly and annual reports

In [19]:
df['quarter_distance'] = df['diff'].where(df['quarterly_report'] == 1, pd.NaT)
df['annual_distance'] = df['diff'].where(df['annual_report'] == 1, pd.NaT)
print(df[['tic', 'rdq', 'fyr', 'fyearq', 'fqtr', 'quarterly_report', 'annual_report', 'fyr_change_dummy', 'market_cap', 'lower_bound', 'upper_bound', 'annual_distance', 'quarter_distance', 'diff']].head())


      tic        rdq  fyr  fyearq  fqtr  quarterly_report  annual_report  \
58  0015B 1998-09-23    5    1998     1                 1              0   
59  0015B 1998-12-23    5    1998     2                 1              0   
60  0015B 1999-03-17    5    1998     3                 1              0   
61  0015B        NaT    5    1998     4                 0              1   
62  0015B 1999-09-15    5    1999     1                 1              0   

    fyr_change_dummy  market_cap  lower_bound  upper_bound annual_distance  \
58                 0  991.896313    793.51705  1190.275575             NaT   
59                 0  719.272500    575.41800   863.127000             NaT   
60                 0  598.121000    478.49680   717.745200             NaT   
61                 0  783.000000    626.40000   939.600000             NaT   
62                 0  765.699000    612.55920   918.838800             NaT   

   quarter_distance     diff  
58         378 days 378 days  
59          

In [21]:
import pandas as pd
import numpy as np

# Initialize counters
empty_quarter_distance_count = 0
empty_annual_distance_count = 0
fyr_change_tickers = df[df['fyr_change_dummy'] == 1]['tic'].unique()

# Iterate over tickers for which fiscal year change happened
for ticker in fyr_change_tickers:
    fyr_change_row = df[(df['tic'] == ticker) & (df['fyr_change_dummy'] == 1)].iloc[0]
    is_quarterly_report = fyr_change_row['quarterly_report'] == 1

    # Filter tickers with similar market capitalization
    similar_cap_tickers = df[(df['market_cap'] >= fyr_change_row['lower_bound']) &
                             (df['market_cap'] <= fyr_change_row['upper_bound'])]

    if is_quarterly_report:
        # For quarterly report, use quarter_distance
        distances = similar_cap_tickers['quarter_distance'].dropna()
        if distances.empty:
            empty_quarter_distance_count += 1
    else:
        # For annual report, use annual_distance
        distances = similar_cap_tickers['annual_distance'].dropna()
        if distances.empty:
            empty_annual_distance_count += 1

print(f"Empty quarter distance count: {empty_quarter_distance_count}")
print(f"Empty annual distance count: {empty_annual_distance_count}")


Empty quarter distance count: 408
Empty annual distance count: 4


In [22]:
import numpy as np
import pandas as pd

# Ensure the 'median_distance' column is created with the correct dtype
df['median_distance'] = np.nan

# Initialize an empty list to store median distances
median_distances = []

# Iterate over tickers where fiscal year change occurred
for ticker in fyr_change_tickers:
    # Step 1: Check if fiscal year change happened in quarterly or annual report
    fyr_change_row = df[(df['tic'] == ticker) & (df['fyr_change_dummy'] == 1)].iloc[0]
    is_quarterly_report = fyr_change_row['quarterly_report'] == 1

    # Step 2: Filter tickers with similar market capitalization
    similar_cap_tickers = df[(df['market_cap'] >= fyr_change_row['lower_bound']) &
                             (df['market_cap'] <= fyr_change_row['upper_bound'])]

    # Step 3: Check if similar cap tickers have non-empty distances
    if is_quarterly_report:
        # For quarterly report, use quarter_distance
        distances = similar_cap_tickers['quarter_distance'].dropna().apply(lambda x: pd.Timedelta(x).days)
    else:
        # For annual report, use annual_distance
        distances = similar_cap_tickers['annual_distance'].dropna().apply(lambda x: pd.Timedelta(x).days)

    # Step 4: Ensure distances list is not empty before calculating median
    if not distances.empty:
        median_distance = distances.median()
    else:
        # If there are no distances, set to NaN
        median_distance = np.nan

    # Step 5: Add the calculated median distance to the DataFrame
    df.loc[(df['tic'] == ticker) & (df['fyr_change_dummy'] == 1), 'median_distance'] = median_distance

# Print the DataFrame to check the updated median_distance column
print(df[['tic', 'rdq', 'fyr', 'fyearq', 'fqtr', 'fyr_change_dummy', 'market_cap', 'lower_bound', 'upper_bound', 'annual_distance', 'quarter_distance', 'median_distance']].head())


      tic        rdq  fyr  fyearq  fqtr  fyr_change_dummy  market_cap  \
58  0015B 1998-09-23    5    1998     1                 0  991.896313   
59  0015B 1998-12-23    5    1998     2                 0  719.272500   
60  0015B 1999-03-17    5    1998     3                 0  598.121000   
61  0015B        NaT    5    1998     4                 0  783.000000   
62  0015B 1999-09-15    5    1999     1                 0  765.699000   

    lower_bound  upper_bound annual_distance quarter_distance  median_distance  
58    793.51705  1190.275575             NaT         378 days              NaN  
59    575.41800   863.127000             NaT          91 days              NaN  
60    478.49680   717.745200             NaT          84 days              NaN  
61    626.40000   939.600000             NaT              NaT              NaN  
62    612.55920   918.838800             NaT              NaT              NaN  


In [23]:
non_empty_median_distances = df['median_distance'].notna().sum()
print(f"Non-empty median distances: {non_empty_median_distances}")

Non-empty median distances: 1634


In [24]:
# Count the actual number of non-empty median distances
actual_non_empty_median_count = df['median_distance'].notna().sum()

# Expected count for quarterly reports
expected_non_empty_median_quarterly = 1911 - 408

# Expected count for annual reports
expected_non_empty_median_annual = 159 - 4

# Print the actual and expected counts
print("Actual non-empty median distances:", actual_non_empty_median_count)
print("Expected non-empty median distances for quarterly reports:", expected_non_empty_median_quarterly)
print("Expected non-empty median distances for annual reports:", expected_non_empty_median_annual)


Actual non-empty median distances: 1634
Expected non-empty median distances for quarterly reports: 1503
Expected non-empty median distances for annual reports: 155


In [25]:
# Ensure the 'median_distance' column is created with the correct dtype
df['median_distance'] = np.nan

# Initialize an empty list to store median distances
median_distances = []

# Iterate over tickers where fiscal year change occurred
for ticker in fyr_change_tickers:
    # Step 1: Check if fiscal year change happened in quarterly or annual report
    fyr_change_row = df[(df['tic'] == ticker) & (df['fyr_change_dummy'] == 1)].iloc[0]
    is_quarterly_report = fyr_change_row['quarterly_report'] == 1

    # Step 2: Filter tickers with similar market capitalization
    similar_cap_tickers = df[(df['market_cap'] >= fyr_change_row['lower_bound']) &
                             (df['market_cap'] <= fyr_change_row['upper_bound'])]

    # Step 3: Check if similar cap tickers have non-empty distances
    if is_quarterly_report:
        # For quarterly report, use quarter_distance
        distances = similar_cap_tickers['quarter_distance'].dropna().apply(lambda x: pd.Timedelta(x).days)
    else:
        # For annual report, use annual_distance
        distances = similar_cap_tickers['annual_distance'].dropna().apply(lambda x: pd.Timedelta(x).days)

    # Step 4: Ensure distances list is not empty before calculating median
    if not distances.empty:
        median_distance = distances.median()
    else:
        # If there are no distances, set to NaN
        median_distance = np.nan

    # Step 5: Add the calculated median distance to the DataFrame
    df.loc[(df['tic'] == ticker) & (df['fyr_change_dummy'] == 1), 'median_distance'] = median_distance

# Print the DataFrame to check the updated median_distance column
print(df[['tic', 'rdq', 'fyr', 'fyearq', 'fqtr', 'fyr_change_dummy', 'market_cap', 'lower_bound', 'upper_bound', 'annual_distance', 'quarter_distance', 'median_distance']].head())


      tic        rdq  fyr  fyearq  fqtr  fyr_change_dummy  market_cap  \
58  0015B 1998-09-23    5    1998     1                 0  991.896313   
59  0015B 1998-12-23    5    1998     2                 0  719.272500   
60  0015B 1999-03-17    5    1998     3                 0  598.121000   
61  0015B        NaT    5    1998     4                 0  783.000000   
62  0015B 1999-09-15    5    1999     1                 0  765.699000   

    lower_bound  upper_bound annual_distance quarter_distance  median_distance  
58    793.51705  1190.275575             NaT         378 days              NaN  
59    575.41800   863.127000             NaT          91 days              NaN  
60    478.49680   717.745200             NaT          84 days              NaN  
61    626.40000   939.600000             NaT              NaT              NaN  
62    612.55920   918.838800             NaT              NaT              NaN  


In [26]:
# Count the total fiscal year changes in quarterly reports
total_quarterly_changes = df[df['quarterly_report'] == 1]['fyr_change_dummy'].sum()

# Count the total fiscal year changes in annual reports
total_annual_changes = df[df['annual_report'] == 1]['fyr_change_dummy'].sum()

print(f"Total fiscal year changes in quarterly reports: {total_quarterly_changes}")
print(f"Total fiscal year changes in annual reports: {total_annual_changes}")


Total fiscal year changes in quarterly reports: 1911
Total fiscal year changes in annual reports: 159


In [27]:
# Initialize counters
empty_quarter_distance_count = 0
empty_annual_distance_count = 0

# Iterate over tickers for which fiscal year change happened
for ticker in fyr_change_tickers:
    fyr_change_row = df[(df['tic'] == ticker) & (df['fyr_change_dummy'] == 1)].iloc[0]
    is_quarterly_report = fyr_change_row['quarterly_report'] == 1

    # Filter tickers with similar market capitalization
    similar_cap_tickers = df[(df['market_cap'] >= fyr_change_row['lower_bound']) &
                             (df['market_cap'] <= fyr_change_row['upper_bound'])]

    if is_quarterly_report:
        # For quarterly report, use quarter_distance
        distances = similar_cap_tickers['quarter_distance'].dropna()
        if distances.empty:
            empty_quarter_distance_count += 1
    else:
        # For annual report, use annual_distance
        distances = similar_cap_tickers['annual_distance'].dropna()
        if distances.empty:
            empty_annual_distance_count += 1

print(f"Empty quarter distance count: {empty_quarter_distance_count}")
print(f"Empty annual distance count: {empty_annual_distance_count}")


Empty quarter distance count: 408
Empty annual distance count: 4


In [29]:
# Count the number of fiscal year changes in quarterly reports and annual reports
fyr_change_quarterly_count = df[df['quarterly_report'] == 1]['fyr_change_dummy'].sum()
fyr_change_annual_count = df[df['annual_report'] == 1]['fyr_change_dummy'].sum()

print("Total fiscal year changes in quarterly reports:", fyr_change_quarterly_count)
print("Total fiscal year changes in annual reports:", fyr_change_annual_count)


Total fiscal year changes in quarterly reports: 1911
Total fiscal year changes in annual reports: 159
